ramips: add support for Mofi 5500

rickmacgillis opened the pull request:
Specifications:

SoC: Mediatek MT7621AT (880 MHz MIPS dual-core, quad-thread, CPU)
512 Megabyte DDR3 SDRAM
32 Megabyte NOR Flash
4 Gigabit RJ45 PoE ports
2 MT7615N wifi chips (2.4GHz and 5GHz)
2 USB ports (1xUSB2 and 1xUSB3 - GL3510 chip)
RJ45 RS232 port on front panel (Max3232 chip)
2x mPCIe 2.0 slots for 4G/5G cards
2x SIM slot
1x SDCard Slot
Power via DC12V
4x Cell Antennae
4x Wifi Antennae

MAC Address Locations:
Purpose	Ex.		Partition	Offset
2.4 Ghz	*:01	factory		0x4
5 GHz	*:02	factory		0x8004
LAN		*:03	factory		0xe000
WAN		*:04	factory		0xe006

MAC address prefix E4:3A:65 is registered to MofiNetwork Inc and used as the prefix for all MAC addresses.

Manual: https://mofinetwork.com/files/MoFi_Network_MOFI5500_5GXeLTE_EM7690_SPECS.pdf
WiFi chip specs: https://www.mediatek.com/products/broadband-wifi/mt7615
CPU chip specs: https://www.mediatek.com/products/home-networking/mt7621

Installation:

Update Mofi 5500 to at least stock firmware version 4.8.6. (Available on the Mofi website.) Previous versions are untested in the upgrade process. Log into the LuCI web interface, usually at 192.168.10.1 and visit the 'System->Backup/Flash Firmware' page. Upload and flash the firmware as usual.

Note to Maintainers: Do not remove SUPPORTED_DEVICES from the Makefile! The customized Mofi version of OpenWRT (stock firmware) expects to see mofi5500 as the device name. The stock firmware does not allow for forcing an installation. Without this line, users cannot upload the new firmware through the stock Mofi firmware.

This device uses cell modems that could use QMI or MBIM. Add LuCI Modem Manager to allow people to use these. Also, if they have two cell network cards, ethernet, USB, or other kinds of networks, they may wish to use MWAN3 to allow failover amongst their networks. Please compile it with mwan3 for multiple WAN connections.

namiltd commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,165 @@
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "green:power";
```
Please use the new LED function definition:
```
    #include <dt-bindings/leds/common.h>
```
```
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
```
The behavior of the other LEDs has not been defined yet.

rickmacgillis replied:
5GHz and 2.4GHz are hardwired in and not available for the firmware. I did update the DTS with the following definition. Once these three comments are sorted, I'll push it. LMK if this is the only change I need to make after reading my other comments.

```
                led_power: power {
			label = "green:power";
                        function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

```

Also, I'm not sure what you mean by this bit.
> They should be defined in linux,default-trigger or in file target\linux\ramips\mt7621\base-files\etc\board.d\01_leds

namiltd replied:
Remove the line `label = "green:power";`

rickmacgillis replied:
Removed locally.

namiltd commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2052,6 +2052,19 @@ define Device/mikrotik_routerboard-m33g
 
+# Do not remove SUPPORTED_DEVICES! The customized Mofi version of OpenWRT (stock firmware) expects
+# to see mofi5500 as the device name. The stock firmware does not allow for forcing an installation.
+# Without this line, users cannot upload the new firmware through the stock Mofi firmware.
```
I think it is unnecessary to post this information here.

rickmacgillis replied:
@bmork asked me to add that in [Issue 13722](https://github.com/openwrt/openwrt/pull/13722#issuecomment-1768661068) It's very important that it stays in there so others don't change it and make it not possible to upload the firmware to their device in the traditional fashion.

rickmacgillis replied:
I reread the comment from @bmork and he said to put it in the commit. I've move this to the commit message, then.

namiltd commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,165 @@
+		};
+
+		internet {
+			label = "green:internet";
```
```
			function = LED_FUNCTION_WAN;
			color = <LED_COLOR_ID_GREEN>;
```

rickmacgillis replied:
Logically that makes sense, but the LEDs on the 5500 are as follows. The ones not in the DTS are ones that are hard wired in, and cannot be controlled by the firmware.

1. Module #1 (This is the cell phone network module in the first PCI-e slot.)
2. Module #2
3. WAN
4. 5GHz WiFi
5. 2.4GHz Wifi
6. Internet
7. Power / Booting

Only the two modules, the "internet" LED, and the power/boot LED are able to be used in the firmware (GPIO), hence why only those exist in the DTSI. I'm looking at my 5500 right now and I see that the "internet" LED blinks at a different interval from WAN and the modules/WiFi LED. It's just occasionally flashing. I searched both the internet and my manual stash to find a 5500 manual, but I couldn't find one. Nothing seems to tell me the difference between "internet" and "WAN". I just mapped it to the same GPIO as the DTB that I decompiled from Mifi IIRC.

namiltd commented:
```
	led_power: led-0 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

		led-1 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <1>;
			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
		};

		led-2 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <2>;
			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
		};

		led-3 {
			function = "internet"; // LED_FUNCTION_INTERNET;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
		};
	};
```

rickmacgillis commented:
@namiltd Thank you for that very complete message. I checked `/etc/config/system` and it does show the `link tx rx` is the correct one. I've pushed the changes.

namiltd commented:
You forgot to push changes in the mt7621_mofinetwork_mofi5500-5gxelte.dts file.

rickmacgillis commented:
Updated.

namiltd commented:
What about ports mac addresses? Are there constants? Where they are defined?

rickmacgillis commented:
I'm not sure what you mean. This is just adding support for the Mofi 5500 device. Are you talking about the switch ports or console port or something? If so, here's an excerpt from the DTS file.

```
&switch0 {
	ports {
		port@0 {
			status = "okay";
			reg = <0>;
			label = "lan1";
		};

		port@1 {
			status = "okay";
			reg = <1>;
			label = "lan2";
		};

		port@2 {
			status = "okay";
			reg = <2>;
			label = "lan3";
		};

		port@3 {
			status = "okay";
			reg = <3>;
			label = "lan4";
		};

		port@4 {
			status = "okay";
			reg = <4>;
			label = "wan";
		};
	};
};
```

namiltd commented:
https://192.168.1.1/cgi-bin/luci/admin/network/network
https://openwrt.org/docs/guide-developer/mac.address

rickmacgillis commented:
Is this necessary for any useful purpose? I mean, OpenWRT automatically found the two wifi macs somewhere, but I can't find them at all. The LAN/WAN mac address is auto-generated. I was hoping to find the wifi macs so I could see if the LAN mac was nearby. It's taking hours, and I'm getting nowhere.

namiltd commented:
Check if these MAC addresses are constant and do not change after each reboot.

rickmacgillis commented:
The Wi-Fi Macs stay the same. The LAN and WAN Macs change.

namiltd commented:
Upload mtd3 (factory) content to namiltd AT yahoo DOT com
Is there a MAC address listed on the device label and what is it?

namiltd commented:
I analyzed MOFI5500_v4_9_5std.bin thanks to binwalk. Partition names should be changed to:
```
 partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};

			factory: partition@40000 {
				label = "factory";
				reg = <0x40000 0x10000>;
				read-only;
			};

			partition@50000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x50000 0x1ab0000>;
			};

			partition@1b00000 {
				label = "Recovery";
				reg = <0x1b00000 0x500000>;
			};
		};
	};
```

rickmacgillis commented:
I updated the partitions locally and they'll be with the next push. MoFi put their mounting screws under the mobile antennae, and it was a pain to get the thing off the wall. I did, however, get it off and found that the only MAC listed is the one for WiFi. I have a whole lot of pictures of all angles of the exterior, and the chips inside. Figured I should grab those while I have the chance.

I'd have to imagine that all Mofi 5500 devices will have relatively unique MAC addresses, as a MAC is important for other devices to know what to talk to at Layer 2. If there were two with the same MAC and different gateway IPs, then presumably both networks would get the same traffic, which further increases the load on both devices necessarily, and potentially opens up security risks. If that's not the case, I'm open to investigating your reasoning.

The reason I bring that up is because the orginal factory firmware from the MoFi website will almost certainly not have a MAC address in it. MoFi probably loads the factory firmware with a shell script that reaches out to one of their servers and asks it for the next MAC address in series, then saves it locally and deletes the script. This way when they manufacture a new device, they need only load the firmware, and let it finish booting up before shipping them out. The firmware on the website will lack that script since their servers aren't public.

namiltd commented:
Fragment of the file /usr/sbin/generic_prof:
```
wifi2_mac=$(mtd_get_mac_binary factory 0x4)
wifi5_mac=$(macaddr_add "$wifi2_mac" 1)

lan_mac=$(macaddr_add "$wifi2_mac" 2)
wan_mac=$(macaddr_add "$wifi2_mac" 3)

uci set -q network.lan_dev.macaddr=$lan_mac
uci set -q network.wan_dev.macaddr=$wan_mac
```

rickmacgillis commented:
What do I add to the DTS for the following? The OpenWRT article ensures that everything is as chaotic and confusing as possible, so I have no idea where to set these. I don't want calculations when the offsets are so sparse. That's spaghetti code. I need my code to be as clean and dependable as possible.

```
MAC Address Locations:
Purpose	Ex.	Partition	Offset
2.4 Ghz	*:01	factory		0x4
5 GHz	*:02	factory		0x8004
LAN	*:03	factory		0xe000
WAN	*:04	factory		0xe006
```

namiltd commented:
Try this:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "mt7621.dtsi"

#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/input/input.h>
#include <dt-bindings/leds/common.h>

/ {
	compatible = "mofinetwork,mofi5500-5gxelte", "mediatek,mt7621-soc";
	model = "MoFi Network MOFI5500-5GXeLTE";

	aliases {
		led-boot = &led_power;
		led-failsafe = &led_power;
		led-running = &led_power;
		led-upgrade = &led_power;
	};

	chosen {
		bootargs = "console=ttyS0,115200";
	};

	leds {
		compatible = "gpio-leds";

		led_power: led-0 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

		led-1 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <1>;
			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
		};

		led-2 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <2>;
			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
		};

		led_internet: led-3 {
			function = LED_FUNCTION_WAN_ONLINE;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
		};
	};

	gpio-keys {
		compatible = "gpio-keys-polled";

		reset {
			label = "reset";
			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
			linux,code = <KEY_RESTART>;
		};
	};
};

&gmac0 {
	nvmem-cells = <&macaddr_factory_e000 0>;
	nvmem-cell-names = "mac-address";
};

&i2c {
	status = "okay";
};

&pcie {
	status = "okay";
};

&pcie0 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		mediatek,mtd-eeprom = <&factory 0x0000>;
		ieee80211-freq-limit = <2400000 2500000>;

		nvmem-cells = <&macaddr_factory_4 0>;
		nvmem-cell-names = "mac-address";
	};
};

&pcie1 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		mediatek,mtd-eeprom = <&factory 0x8000>;
		ieee80211-freq-limit = <5000000 6000000>;

		nvmem-cells = <&macaddr_factory_8004 0>;
		nvmem-cell-names = "mac-address";
	};
};

&sdhci {
	status = "okay";
};

&spi0 {
	status = "okay";

	m25p80@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <50000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};

			factory: partition@40000 {
				label = "factory";
				reg = <0x40000 0x10000>;
				read-only;

				nvmem-layout {
					compatible = "fixed-layout";
					#address-cells = <1>;
					#size-cells = <1>;

					macaddr_factory_4: macaddr@4 {
						reg = <0x4 0x6>;
					};

					macaddr_factory_8004: macaddr@8004 {
						reg = <0x8004 0x6>;
					};

					macaddr_factory_e000: macaddr@e000 {
						reg = <0xe000 0x6>;
					};

					macaddr_factory_e006: macaddr@e006 {
						reg = <0xe006 0x6>;
					};

				);

			};

			partition@50000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x50000 0x1ab0000>;
			};

			partition@1b00000 {
				label = "Recovery";
				reg = <0x1b00000 0x500000>;
			};
		};
	};
};

&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan1";
		};

		port@1 {
			status = "okay";
			label = "lan2";
		};

		port@2 {
			status = "okay";
			label = "lan3";
		};

		port@3 {
			status = "okay";
			label = "lan4";
		};

		port@4 {
			status = "okay";
			label = "wan";

			nvmem-cells = <&macaddr_factory_e006 0>;
			nvmem-cell-names = "mac-address";
		};
	};
};
```

namiltd commented:
Try this:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "mt7621.dtsi"

#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/input/input.h>
#include <dt-bindings/leds/common.h>

/ {
	compatible = "mofinetwork,mofi5500-5gxelte", "mediatek,mt7621-soc";
	model = "MoFi Network MOFI5500-5GXeLTE";

	aliases {
		led-boot = &led_power;
		led-failsafe = &led_power;
		led-running = &led_power;
		led-upgrade = &led_power;
	};

	chosen {
		bootargs = "console=ttyS0,115200";
	};

	leds {
		compatible = "gpio-leds";

		led_power: led-0 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

		led-1 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <1>;
			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
		};

		led-2 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <2>;
			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
		};

		led_internet: led-3 {
			function = LED_FUNCTION_WAN_ONLINE;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
		};
	};

	gpio-keys {
		compatible = "gpio-keys-polled";

		reset {
			label = "reset";
			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
			linux,code = <KEY_RESTART>;
		};
	};
};

&gmac0 {
	nvmem-cells = <&macaddr_factory_e000 0>;
	nvmem-cell-names = "mac-address";
};

&i2c {
	status = "okay";
};

&pcie {
	status = "okay";
};

&pcie0 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <2400000 2500000>;

		nvmem-cells = <&eeprom_factory_0>, <&macaddr_factory_4 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&pcie1 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <5000000 6000000>;

		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_8004 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&sdhci {
	status = "okay";
};

&spi0 {
	status = "okay";

	m25p80@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <50000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};

			factory: partition@40000 {
				label = "factory";
				reg = <0x40000 0x10000>;
				read-only;

				nvmem-layout {
					compatible = "fixed-layout";
					#address-cells = <1>;
					#size-cells = <1>;

					eeprom_factory_0: eeprom@0 {
						reg = <0x0 0x6000>;
					};

					macaddr_factory_4: macaddr@4 {
						reg = <0x4 0x6>;
					};

					eeprom_factory_8000: eeprom@8000 {
						reg = <0x8000 0x6000>;
					};

					macaddr_factory_8004: macaddr@8004 {
						reg = <0x8004 0x6>;
					};

					macaddr_factory_e000: macaddr@e000 {
						reg = <0xe000 0x6>;
					};

					macaddr_factory_e006: macaddr@e006 {
						reg = <0xe006 0x6>;
					};

				);

			};

			partition@50000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x50000 0x1ab0000>;
			};

			partition@1b00000 {
				label = "Recovery";
				reg = <0x1b00000 0x500000>;
			};
		};
	};
};

&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan1";
		};

		port@1 {
			status = "okay";
			label = "lan2";
		};

		port@2 {
			status = "okay";
			label = "lan3";
		};

		port@3 {
			status = "okay";
			label = "lan4";
		};

		port@4 {
			status = "okay";
			label = "wan";

			nvmem-cells = <&macaddr_factory_e006 0>;
			nvmem-cell-names = "mac-address";
		};
	};
};
```
The values ​​for eeprom_factory_0 and eeprom_factory_8000 need to be verified.

namiltd commented:
Try this:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "mt7621.dtsi"

#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/input/input.h>
#include <dt-bindings/leds/common.h>

/ {
	compatible = "mofinetwork,mofi5500-5gxelte", "mediatek,mt7621-soc";
	model = "MoFi Network MOFI5500-5GXeLTE";

	aliases {
		led-boot = &led_power;
		led-failsafe = &led_power;
		led-running = &led_power;
		led-upgrade = &led_power;
	};

	chosen {
		bootargs = "console=ttyS0,115200";
	};

	leds {
		compatible = "gpio-leds";

		led_power: led-0 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

		led-1 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <1>;
			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
		};

		led-2 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <2>;
			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
		};

		led_internet: led-3 {
			function = LED_FUNCTION_WAN_ONLINE;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
		};
	};

	gpio-keys {
		compatible = "gpio-keys-polled";

		reset {
			label = "reset";
			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
			linux,code = <KEY_RESTART>;
		};
	};
};

&gmac0 {
	nvmem-cells = <&macaddr_factory_e000 0>;
	nvmem-cell-names = "mac-address";
};

&i2c {
	status = "okay";
};

&pcie {
	status = "okay";
};

&pcie0 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <2400000 2500000>;

		nvmem-cells = <&eeprom_factory_0>, <&macaddr_factory_4 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&pcie1 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <5000000 6000000>;

		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_8004 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&sdhci {
	status = "okay";
};

&spi0 {
	status = "okay";

	m25p80@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <50000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};

			factory: partition@40000 {
				label = "factory";
				reg = <0x40000 0x10000>;
				read-only;

				nvmem-layout {
					compatible = "fixed-layout";
					#address-cells = <1>;
					#size-cells = <1>;

					eeprom_factory_0: eeprom@0 {
						reg = <0x0 0x4da8>;
					};

					macaddr_factory_4: macaddr@4 {
						reg = <0x4 0x6>;
					};

					eeprom_factory_8000: eeprom@8000 {
						reg = <0x8000 0x4da8>;
					};

					macaddr_factory_8004: macaddr@8004 {
						reg = <0x8004 0x6>;
					};

					macaddr_factory_e000: macaddr@e000 {
						reg = <0xe000 0x6>;
					};

					macaddr_factory_e006: macaddr@e006 {
						reg = <0xe006 0x6>;
					};

				);

			};

			partition@50000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x50000 0x1ab0000>;
			};

			partition@1b00000 {
				label = "Recovery";
				reg = <0x1b00000 0x500000>;
			};
		};
	};
};

&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan1";
		};

		port@1 {
			status = "okay";
			label = "lan2";
		};

		port@2 {
			status = "okay";
			label = "lan3";
		};

		port@3 {
			status = "okay";
			label = "lan4";
		};

		port@4 {
			status = "okay";
			label = "wan";

			nvmem-cells = <&macaddr_factory_e006 0>;
			nvmem-cell-names = "mac-address";
		};
	};
};
```
The values ​​for eeprom_factory_0 and eeprom_factory_8000 need to be verified. Not sure what the size is for the MediaTek MT7615N.

Of course, analyze everything because I may have simple typos.

rickmacgillis commented:
Thank you. I pushed it.

rickmacgillis commented:
Updated.

namiltd commented:
"Co-authored-by:" should be on the penultimate line (before Signed-off-by:).

rickmacgillis commented:
I'll update it on the next commit since placement is just fluff. Do we have anything else to tackle?

namiltd commented:
Test it on hardware.

rickmacgillis commented:
I can confirm that the WAN/LAN Mac addresses are correct, and I can access the internet. (I'm connected through my router right now on the new firmware.) However, `Network > Wireless` is missing, and I'm trying to find out where in `make menuconfig` I need to add that in.

Do you know where I can see a list of things that OpenWRT releases usually have baked right in? Maybe there's a way to auto-select those on my build.

namiltd commented:
Add: `$(Device/dsa-migration)`

rickmacgillis commented:
Where do I put `$(Device/dsa-migration)`?

namiltd commented:
target/linux/ramips/image/mt7621.mk line #2056

namiltd commented:
Add to target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts file:
```
&state_default {
	gpio {
		groups = "jtag", "wdt";
		function = "gpio";
	};
};
```

namiltd commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,225 @@
+	};
+};
+
+&state_default {
```
Move it down to make it alphabetical

namiltd commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2052,6 +2052,17 @@ define Device/mikrotik_routerboard-m33g
 
+define Device/mofinetwork_mofi5500-5gxelte
+  $(Device/uimage-lzma-loader)
+  $(Device/dsa-migration)
```
Move it up

namiltd commented:
Add missing label-mac-device:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "mt7621.dtsi"

#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/input/input.h>
#include <dt-bindings/leds/common.h>

/ {
	compatible = "mofinetwork,mofi5500-5gxelte", "mediatek,mt7621-soc";
	model = "MoFi Network MOFI5500-5GXeLTE";

	aliases {
		label-mac-device = &wifi0;
		led-boot = &led_power;
		led-failsafe = &led_power;
		led-running = &led_power;
		led-upgrade = &led_power;
	};

	chosen {
		bootargs = "console=ttyS0,115200";
	};

	leds {
		compatible = "gpio-leds";

		led_power: led-0 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};

		led-1 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <1>;
			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
		};

		led-2 {
			function = LED_FUNCTION_MOBILE;
			color = <LED_COLOR_ID_BLUE>;
			function-enumerator = <2>;
			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
		};

		led_internet: led-3 {
			function = LED_FUNCTION_WAN_ONLINE;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
		};
	};

	gpio-keys {
		compatible = "gpio-keys-polled";

		reset {
			label = "reset";
			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
			linux,code = <KEY_RESTART>;
		};
	};
};

&gmac0 {
	nvmem-cells = <&macaddr_factory_e000 0>;
	nvmem-cell-names = "mac-address";
};

&i2c {
	status = "okay";
};

&pcie {
	status = "okay";
};

&pcie0 {
	wifi0: wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <2400000 2500000>;

		nvmem-cells = <&eeprom_factory_0>, <&macaddr_factory_4 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&pcie1 {
	wifi1: wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <5000000 6000000>;

		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_8004 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};

&sdhci {
	status = "okay";
};

&spi0 {
	status = "okay";

	m25p80@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <50000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};

			factory: partition@40000 {
				label = "factory";
				reg = <0x40000 0x10000>;
				read-only;

				nvmem-layout {
					compatible = "fixed-layout";
					#address-cells = <1>;
					#size-cells = <1>;

					eeprom_factory_0: eeprom@0 {
						reg = <0x0 0x4da8>;
					};

					macaddr_factory_4: macaddr@4 {
						compatible = "mac-base";
						reg = <0x4 0x6>;
						#nvmem-cell-cells = <1>;
					};

					eeprom_factory_8000: eeprom@8000 {
						reg = <0x8000 0x4da8>;
					};

					macaddr_factory_8004: macaddr@8004 {
						compatible = "mac-base";
						reg = <0x8004 0x6>;
						#nvmem-cell-cells = <1>;
					};

					macaddr_factory_e000: macaddr@e000 {
						compatible = "mac-base";
						reg = <0xe000 0x6>;
						#nvmem-cell-cells = <1>;
					};

					macaddr_factory_e006: macaddr@e006 {
						compatible = "mac-base";
						reg = <0xe006 0x6>;
						#nvmem-cell-cells = <1>;
					};

				};

			};

			partition@50000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x50000 0x1ab0000>;
			};

			partition@1b00000 {
				label = "Recovery";
				reg = <0x1b00000 0x500000>;
			};
		};
	};
};

&state_default {
	gpio {
		groups = "jtag", "wdt";
		function = "gpio";
	};
};

&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan1";
		};

		port@1 {
			status = "okay";
			label = "lan2";
		};

		port@2 {
			status = "okay";
			label = "lan3";
		};

		port@3 {
			status = "okay";
			label = "lan4";
		};

		port@4 {
			status = "okay";
			label = "wan";

			nvmem-cells = <&macaddr_factory_e006 0>;
			nvmem-cell-names = "mac-address";
		};
	};
};
```

namiltd commented:
You missed the changes in lines 80 and 91.

namiltd commented:
Regarding Modem manager idea I also have no experience.

rickmacgillis commented:
> You missed the changes in lines 80 and 91.

I've updated those locally. Why does the label only show `wifi0` (2.4GHz) and not also specify the `wifi1` label?

As for the Modem Manager stuff, let's drop it for now. If someone else reviewing this PR wants to chime in, that'll be helpful.

rickmacgillis commented:
Updated. Ready for the approvals stage, or is there anything left to do? Any point in flashing the firmware again, or is that good to go since the builds below will succeed?

namiltd commented:
Flashing is not needed because it is a minor change.

There is still no ' wifi0: wifi@0,0 {' on line 80 and ` wifi1: wifi@0,0 {' on line 91

I have no other comments regarding the code.

rickmacgillis commented:
Pushed again. Please give it an approving review.

rickmacgillis commented:
Tagging a few members I have interacted with in the past to see if we can get the approvals stage finalized.

@PolynomialDivision
@raenye
@musashino205 
@mrkiko 
@bmork 
@hnyman

namiltd commented:
We can check the code, but it's very hard to check the functionality if we don't have this hardware. At least include a kernel log.

rickmacgillis commented:
Kernel Log can be seen at https://gist.github.com/rickmacgillis/bac097e3c54389bc37390aa6958764e6

namiltd commented:
To avoid the error "OF: Invalid cell count for /palmbus@1e000000/spi@b00/m25p80@0/partitions" change:
```
&spi0 {
	status = "okay";

	m25p80@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		#address-cells = <1>;
		#size-cells = <1>;
		spi-max-frequency = <50000000>;

		partitions {
```

rickmacgillis commented:
Updated. Anything else we need to change?

namiltd commented:
Now we are waiting for comments from others.

rickmacgillis commented:
I see this in the logs, but the DTS doesn't specify the USB, so it's probably coming from the DTSI. Is this something we need to handle?

```
[    5.705509] usb 1-1.4: config 1 has an invalid interface number: 12 but max is 4
[    5.712951] usb 1-1.4: config 1 has an invalid interface number: 13 but max is 4
[    5.720425] usb 1-1.4: config 1 has an invalid interface number: 13 but max is 4
[    5.727886] usb 1-1.4: config 1 has no interface number 1
[    5.733289] usb 1-1.4: config 1 has no interface number 4
[    5.853798] usb 2-1.3: new SuperSpeed USB device number 3 using xhci-mtk
[    5.895022] usb 2-1.3: config 1 has an invalid interface number: 8 but max is 3
[    5.902373] usb 2-1.3: config 1 has no interface number 1
```

raenye commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2052,6 +2052,17 @@ define Device/mikrotik_routerboard-m33g
+  DEVICE_VENDOR := MoFi Network
+  DEVICE_MODEL := MOFI5500-5GXeLTE
+  DEVICE_PACKAGES := kmod-usb3 kmod-usb2 kmod-sdhci-mt7620 kmod-mt7615-firmware
+  SUPPORTED_DEVICES += mofi5500
```
I'd add here a quick comment about the necessity of this (maybe even just "see commit for explanation").

rickmacgillis replied:
Locally I've changed it to `SUPPORTED_DEVICES += mofi5500 # Needed in order to flash through Mofi stock firmware` and I'll push it with the rest of the changes.

raenye commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys-polled";
```
is this necessary? (i.e., do buttons fail to work with `gpio-keys` only)

rickmacgillis replied:
This keeps coming up. I'll test it.

raenye commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+		led-1 {
+			function = LED_FUNCTION_MOBILE;
+			color = <LED_COLOR_ID_BLUE>;
+			function-enumerator = <1>;
```
Enumerate them as 1 and 2, or 0 and 1?
I'd make this match either markings on the case, or entries in `/dev/`.

rickmacgillis replied:
The case says `Module #1` and `Module #2`. Should it stay as-is in the DTS?

raenye replied:
Maybe just add comments in the DTS (e.g. "case says Module #1" and "case says Module #2").

rickmacgillis replied:
Added. I'll push it with the changes.

raenye commented:
Looks good to me.
What's in the recovery partition, BTW?

raenye commented:
Also, maybe change commit text to say "mPCIe slots for 4G/5G cards". Is this PCIe 2.0?

rickmacgillis commented:
I opened the downloaded Recovery partition in a hex editor. The top of it says, `MofiRecovery5500-v1.0-RECOVERY` so I believe it's something that Mofi added as a stock partition.

How do I tell if it's PCI 2.0?

namiltd commented:
For gpio-leds-polled you also need to define poll-interval.

rickmacgillis commented:
@raenye If everything looks good now, please add an approving review.

raenye reviewed (approved):
(here you go, but AFAIK this means the same for committers as writing LGTM in the comments)

rickmacgillis commented:
@hnyman Thank you so much for your help on the forums over at https://forum.openwrt.org/t/what-do-i-need-to-do-to-add-mofi-5500-as-a-target/170437 as it really helped me a lot to build the DTS when I was working on it. I'd really appreciate your insights on this PR and if you think it's ready for your approving review.

namiltd commented:
I extracted the dts file from MOFI5500_v4_9_5std.bin: [mofi.dts.txt](https://github.com/user-attachments/files/19040310/mofi.dts.txt)
You have omitted the mofi:green:modem1_pwr and mofi:green:modem2_pwr LEDs

raenye commented:
These are `led-1` and `led-2`, no?

rickmacgillis commented:
LED-1 and LED-2 are blue LEDs on the face of the router. There are no other external LEDs for the modems. Also, I checked the pictures and I don't readily see any internal LEDs for them, but if your extracted file shows the 2 blue modem LEDs as well, I'm willing to compile that in in case there are internal power LEDs I don't see. It might help someone some day. I don't know the GPIO numbers for those, though. Didn't even know they existed.

Please confirm. Also, if you see anything else missing, LMK.

rickmacgillis commented:
I went through them and it seems that all is there aside from these LED differences. I came up with the following items for the module power LEDs. However, the `dt-bindings/leds/common.h` file doesn't give me any better function constant to name these. I REALLY don't like that everything is `led-1` `led-2` `led-3` and so forth, as that's nondescriptive spaghetti code. With the `function` set to the description of what it does, it helps describe them. I get that `led-1` and others are done for enumeration and iteration, but so others know why they exist, what do I put for the `function` on these LED specifications?

```
                led-4 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
		};

		led-5 {
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_GREEN>;
			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
		};
```

namiltd commented:
It is possible, however, that this firmware is universal for different versions of the router. If there are no other physical LEDs, then there is no issue.

rickmacgillis commented:
While I don't think it's universal since they mentioned the 5500 in the DTS, I'm not sure if it's useful or not. There may be hidden LEDs inside that they reference, but I'm not 100% sure. I guess the best way to do it is just to leave it out. And if somebody later needs those, then we can add them back in.

namiltd commented:
e.g.:
```		function = "mobile_power";
		function-enumerator = <1>; // Case says Module #1
```

mrkiko commented:
Naive question here - have you tried to toggle these GPIOs?
May those two LEDs be related to modules power supply instead?



On Sat, Mar 01, 2025 at 02:07:24PM -0800, Mieczyslaw Nalewaj wrote:
> namiltd left a comment (openwrt/openwrt#18076)
> 
> e.g.:
> ```		function = "mobile_power";
> 		function-enumerator = <1>; // Case says Module #1
> ```
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/18076#issuecomment-2692442208
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

raenye commented:
> have you tried to toggle these GPIOs?

(@rickmacgillis and you don't even have to recompile, just use the `/sys/kernel/debug/gpio/` interface)

rickmacgillis commented:
I just tested it (after adjusting 520 to be the gpio + the output of `cat /sys/class/gpio/gpiochip*/base | head -n1`) and it doesn't change any of the external LEDs. I looked through the vents the best I could and I didn't see any lights change. There's a toggle switch on the case that says "LEDs on/off", and I tried shutting it off to see if anything inside would turn on, and nothing did.

I think either I'm missing it or it's the power supply for the modules. It takes me a good 15 minutes to take the Mofi off the wall, so I rather not do that again. If it is an LED, it's either under the thermal gel pads or the metal shields that cover certain components.

namiltd commented:
There are no additional LEDs so the code does not require any modifications.

rickmacgillis commented:
Do we need to describe the GPIO power supplies in DTS? I'm not sure why they're in there, unless Mofi used it for their software or something.

raenye commented:
If you're not sure they control power, I wouldn't add anything.
If they do, maybe add regulators (see, e.g., [here](https://github.com/u-boot/u-boot/blob/master/arch/arm/dts/imx53-ppd.dts)), but it's better to hear the opinion of a committer.

rickmacgillis commented:
I don't know for sure what Mofi did. I don't know how to check if we need to specify these either. @PolynomialDivision, care to chime in on this one and also about the LuCI Modem Manager question above? These are closely related, as right now there's no way to see the modules (modems), nor validate any power needs.

FYI, my current build shows the modems and sim cards in Modem Manager. I just didn't force Modern Manager in the build since not everyone will use LuCI. Please advise.

rickmacgillis commented:
@mrkiko If you like what you see, please give it an approving review.

rickmacgillis commented:
Hi guys. I'm just checking in to see what we can do to get this merged.

namiltd commented:
Be patient.

rickmacgillis commented:
How long does it usually take?

namiltd commented:
I wait half a year with some of my PRs, but it usually closes in a few days or weeks. If someone else installed it on real hardware and confirmed it works, it would be easier.

rickmacgillis commented:
Thank you. I'll post on the OpenWRT boards to see if someone can beta test it and report back.

mrkiko commented:
Hello!

I don't know a way to approve the PR from CLI.

That said, some notes:
- is there any reason why you prefer to use the mt7620 SDHCI module instead of kmod-mmc-mtk ?
- in the commit description, you mention the SD card slot twice, so you may remove a reference

A question - is it possible to interact with the bootloader via the RJ45 RS232 port?

Thanks a lot for your work.

namiltd commented:
Try: 
```
DEVICE_PACKAGES := kmod-mt7615-firmware kmod-usb3  kmod-mmc-mtk \
             kmod-usb-net-qmi-wwan kmod-usb-net-cdc-mbim  -uboot-envtools
```

mrkiko commented:
Why include uboot-envtools, since there is no config entry added for it?
Or am I missing something?

hauke commented:
> > Why include uboot-envtools, since there is no config entry added for it?
> 
> I originally called the partition `uboot`, but @namiltd renamed it to `Config`. It's been renamed time and again since I made the first release over a year ago. Mofi calls it `Config` and OpenWRT calls it `uboot`. Also, IIRC there was something under the config partition definition for uboot, but now it's gone and because everyone says to squash my commits, I've now lost that history...

The name of the partition does not matter. To make the u-boot env tools work you need an entry in the `package/boot/uboot-envtools/files/ramips` file. 

> > in the commit description, you mention the SD card slot twice, so you may remove a reference
> 
> Removed. I'll use the new commit message once I push any new changed.

kmod-mmc-mtk is the upstream driver, kmod-sdhci-mt7620 is an older driver for the sdcrad controller. Please use `kmod-mmc-mtk`.

rickmacgillis commented:
@hauke Thank you. I've replaced `kmod-sdhci-mt7620` with `kmod-mmc-mtk` in the makefile. I've also added the entry for uboot. Did I do it right?

```
mofinetwork,mofi5500-5gxelte)
	ubootenv_add_uci_config "/dev/mtd0" "0x0" "0x10000" "0x30000"
	;;
```
The non-descript uci variables that the uboot-envtools.sh shell script sets are difficult to understand. Here's the format for the `ubootenv_add_uci_config` command, but I'm not sure if my partition data is correct.

```
ubootenv_add_uci_config [Device Name] [offset] [envsize] [secsize] [numsec]
```
The `numsec` is apparently optional. I looked at other entries and tried to match things up in the DTS for those entries, but  `envsize` doesn't seem to correspond to anything.

Here's my config partition setup for simplicity.

```C
			partition@0 {
				label = "Bootloader";
				reg = <0x0 0x30000>;
				read-only;
			};

			partition@30000 {
				label = "Config";
				reg = <0x30000 0x10000>;
			};
```

raenye commented:
> Yes. However, for some reason using a console cable and running `screen /dev/ttyUSB0 115200` seems to give partly jumbled output. I've also tried 9600 baud, but that doesn't show any output. It works just fine on the stock firmware, though. I'd really like to know if there's a solution to this.

I had a weird UART behaviour with DAP-1720 (first 125000, then 115200); see e29f4a3f70d278c.

(See also my solution for the partition naming question in the same commit)

rickmacgillis commented:
Hi @raenye. I'm not sure what you're trying to show me. What did you change to get better results?

rickmacgillis commented:
So we can stay on track, Ill list what we're still waiting on responses to.

1. Is my `ubootenv_add_uci_config` command, in the comment above, setting things correctly?
2. I'm now running my new firmware, but `sudo screen /dev/ttyUSB0 115200` still gives me jumbled output on boot like the text below. Usually that means it's a different baud rate, but no other baud rate seems to show it properly. It's hit or miss if I can connect to it AFTER it's booted, too.

```
27.76] br-lan: port n3)nteredisabled state   27.752449] mt53iocast  27.763897] mt7530-mdio mdio-bu:1f lan3: enred promiscuous me
```

rickmacgillis commented:
Any updates?

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+};
+
+&i2c {
+	status = "okay";
```
I2C is unused on this model? I guess this node can be removed.

rickmacgillis replied:
How do you know if I2C is used on it or not?

DragonBluep replied:
I don't see I2C peripheral from your commit message description.

rickmacgillis replied:
Is there some equivalent of `lsi2c`? Here's the output from `dmesg`.

```
[root@MOFI5500:/root]#dmesg | grep i2c
[    1.711368] i2c_dev: i2c /dev entries driver
[    1.716618] i2c-mt7621 1e000900.i2c: clock 100 kHz
```

Any better way to do this?

DragonBluep replied:
Maybe you can try using `i2c-tools` to detect if there is an i2c device.

rickmacgillis replied:
I installed it and used the guide at https://gist.github.com/karantir/c3b123a63696a895cc9dad32168fe690

Here's the output of some of the commands:

```
[root@MOFI5500:/var/log]#i2cdetect -y 0
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: 60 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
[root@MOFI5500:/var/log]#i2cdump -y 0 0x60
No size specified (using byte-data access)
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: XX XX XX 04 11 33 43 04 11 33 43 04 11 33 43 04    XXX??3C??3C??3C?
10: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
20: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
30: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
40: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
50: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
60: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
70: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
80: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
90: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
a0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
b0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
c0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
d0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
e0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
f0: 11 33 43 04 11 33 43 04 11 33 43 04 11 33 43 04    ?3C??3C??3C??3C?
```

I also read https://linux.die.net/man/8/i2cget and didn't understand it fully, so I didn't try running the `i2cget` commands, so as to not accidentally mess things up. If you need me to run a read command on it, please send me the command, and I'll run it.

DragonBluep replied:
![image](https://github.com/user-attachments/assets/c9151954-839f-4411-879d-610e733e01c5)
This chip looks like an i2c eeprom chip? ref: https://fccid.io/2AE6X-MOFI5500/Internal-Photos/Internal-Photos-5591739

rickmacgillis replied:
Probably is. I checked my pictures, but the one I took has too much dust on top of that chip to read its number well enough. So, let's keep I2C enabled in case it's used somewhere. It doesn't seem to be interfering with anything else that might use the same pins. (Sometimes, such as with Arduino, circuits have mutually exclusive interfaces that rely on the same pins. Like I2C vs GPIO, but I don't see any problems with it.) If you also agree, I'll mark this as resolved.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
```suggestion
	flash@0 {
```

rickmacgillis replied:
Changed, and I'm not sure why that chip was specified before.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
```
Please don't add `#address-cells` and `#size-cells` here. Wrong syntax.

rickmacgillis replied:
Removed.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+				reg = <0x30000 0x10000>;
+			};
+
+			factory: partition@40000 {
```
```suggestion
			partition@40000 {
```

rickmacgillis replied:
This label seems to be used all over the DTS. Is there a compelling reason to remove it?

DragonBluep replied:
No. It's unused. I don't see `&factory` reference here.

rickmacgillis replied:
Alright. I'll remove it and test. It has a `label` item on that object anyway, so I'm inclined to believe that other things referencing `factory` will still work.

raenye replied:
@rickmacgillis `macaddr_factory_e006` (and friends) are separate labels, and references are to these.

rickmacgillis replied:
Seems fine.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+
+		port@4 {
+			status = "okay";
+			label = "wan";
```
Please mux wan port to GMAC1. ref:
https://git.openwrt.org/openwrt/openwrt.git?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=grep&s=gmac1

rickmacgillis replied:
Is this what you mean? It's already in there.

```
&gmac0 {
	nvmem-cells = <&macaddr_factory_e000 0>;
	nvmem-cell-names = "mac-address";
};
```

Also, notice the `factory` label on it.

rickmacgillis replied:
So you're saying to add the following:

```
&gmac1 {
	status = "okay";
	label = "wan";
	phy-handle = <&ethphy4>;

	nvmem-cells = <&macaddr_factory_e006 0>;
	nvmem-cell-names = "mac-address";
};
```

Then remove the `port@4` section on the list of `ports` for `&switch0`. Correct?

DragonBluep replied:
Yes.

rickmacgillis replied:
Adjusted and lined up for testing.

rickmacgillis replied:
This one failed. I can no longer use the WAN port on my router. I'm talking to you through my hotspot instead. I'm going to push my code, but I don't know what's wrong. I uploaded my archived backup and it still doesn't resolve WAN being offline. Presumably it's because of `phy-handle = <&ethphy4>;` as it was copied from somewhere and I don't know what it's supposed to be. I'll remove it and test to see if that gets WAN working again, but I don't know if that handle is important.

rickmacgillis replied:
Code pushed, but it doesn't have WAN access.

DragonBluep replied:
You also need to add
```
&ethphy4 {
	/delete-property/ interrupts;
};
```

> Is this important to move for a technological reason, or should I push the code with `port@4`?

In this way, you can get double bandwidth from switch to CPU. https://github.com/openwrt/openwrt/pull/10238

rickmacgillis replied:
Thank you, so you're saying to put it back to `gmac1` and additionally add the following. (As opposed to using `port@4` and using the following code.) Is that correct?

```
&ethphy4 {
	/delete-property/ interrupts;
};
```

DragonBluep replied:
Yes. Please see other mt7621 dts.

rickmacgillis replied:
Compiling and testing now.

rickmacgillis replied:
That worked. Code pushed.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2052,6 +2052,18 @@ define Device/mikrotik_routerboard-m33g
+  IMAGE_SIZE := 27656k
+  DEVICE_VENDOR := MoFi Network
+  DEVICE_MODEL := MOFI5500-5GXeLTE
+  DEVICE_PACKAGES := kmod-usb3 kmod-usb2 kmod-sdhci-mt7620 kmod-mt7615-firmware \
```
Use `kmod-mmc-mtk` instead of `kmod-sdhci-mt7620` and remove `kmod-usb2`.

rickmacgillis replied:
This is the second time I've seen this comment. Does `kmod-usb3` also have backwards USB 2.0 compatibility? There are USB 2.0 and 3.0 ports on the device. Also, why are we removing the specific firmware version of MMC support to replace it with the generic Mediatek version?

DragonBluep replied:
> This is the second time I've seen this comment. Does `kmod-usb3` also have backwards USB 2.0 compatibility?

Yep.

rickmacgillis replied:
`kmod-usb2` removed. It seems that the local copy already uses `kmod-mmc-mtk` instead of `kmod-sdhci-mt7620`.

DragonBluep commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -164,6 +164,9 @@ mikrotik,routerboard-m11g)
 	ucidef_set_led_rssi "rssiveryhigh" "RSSIVERYHIGH" "green:rssi4" "wlan0" "80" "100"
 	;;
+mofinetwork,mofi5500-5gxelte)
+	ucidef_set_led_netdev "internet" "Internet" "green:wan-online" "wan" "link tx rx"
```
```suggestion
	ucidef_set_led_netdev "internet" "Internet" "green:wan-online" "wan"
```

rickmacgillis replied:
Whats the reason for removing "link rx tx"? @namiltd, care to chime in here?

DragonBluep replied:
`"link rx tx"` is the default behavior. Just align with other devices.

rickmacgillis replied:
Removed.

rickmacgillis replied:
I think I might have added that to UCI on my old OpenWRT firmware, then checked and it showed that data. If this is the default, then it makes sense to remove it from being explicitly set. I'll test that as well without the "link rx tx" and see if LuCI shows those as checked.

rickmacgillis replied:
Tested and I can confirm that removing these still allows the default link, rx, and tx to be checked.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+	wifi0: wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		ieee80211-freq-limit = <2400000 2500000>;
```
> 2 MT7615N wifi chips (2.4GHz and 5GHz)

The WiFi doesn't work, right? Please attach the bootlog here.

rickmacgillis replied:
How does it not work right?

rickmacgillis replied:
Do you want the syslog or kernlog?

DragonBluep replied:
Any one is fine.

rickmacgillis replied:
Here's a Gist for the syslog. I'll remove it once you've looked at it. https://gist.github.com/rickmacgillis/5465b52d420cfab43cb97d6bb5d57af8

DragonBluep replied:
Strange. This model has 4 PCIe slot, but MT7621 should only has 3. Just forget this comment, the WiFi seems work.

rickmacgillis replied:
Sounds good. Gist deleted. Marking it as resolved.

DragonBluep reviewed (commented):
Please test if `reboot` command works on this device.

rickmacgillis commented:
Yes, the `reboot` command works.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,228 @@
+		ieee80211-freq-limit = <5000000 6000000>;
+
+		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_8004 0>;
+		nvmem-cell-names = "eeprom", "mac-address";
```
This can be simplified to
```
		nvmem-cells = <&eeprom_factory_8000>;
		nvmem-cell-names = "eeprom";
```
offset 4 is the default MAC address for Mediatek NICs.

rickmacgillis replied:
Updated. Does this one also change in the same way?

```
&pcie0 {
	wifi0: wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		ieee80211-freq-limit = <2400000 2500000>;

		nvmem-cells = <&eeprom_factory_0>, <&macaddr_factory_4 0>;
		nvmem-cell-names = "eeprom", "mac-address";
	};
};
```

rickmacgillis replied:
2.4GHz is the one listed on the label.

rickmacgillis replied:
Here are my notes:

```
2.4: E4:3A:65:00:00:01 (0x4)
5:   E4:3A:65:00:00:02 (0x8004)
LAN: E4:3A:65:00:00:03 (0xe000)
WAN: E4:3A:65:00:00:04(0xe006)
```

I hid the last 3 octets for privacy, but this shows the order they come in.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,226 @@
+};
+
+&pcie1 {
+	wifi1: wifi@0,0 {
```
```suggestion
	wifi@0,0 {
```

rickmacgillis replied:
Removed.

DragonBluep commented on target/linux/ramips/dts/mt7621_mofinetwork_mofi5500-5gxelte.dts:
```diff
@@ -0,0 +1,226 @@
+
+					macaddr_factory_8004: macaddr@8004 {
+						compatible = "mac-base";
+						reg = <0x8004 0x6>;
```
Unused and can be removed.

rickmacgillis replied:
Which part? GitHub annoyingly selects multiple lines and I have no idea which one people are talking about half the time.

DragonBluep replied:
```
					macaddr_factory_8004: macaddr@8004 {
						compatible = "mac-base";
						reg = <0x8004 0x6>;
						#nvmem-cell-cells = <1>;
					};
```

rickmacgillis replied:
Removed.

Commit:
ramips: add support for Mofi 5500

Specifications:

SoC: Mediatek MT7621AT (880 MHz MIPS dual-core, quad-thread, CPU)
512 Megabyte DDR3 SDRAM
32 Megabyte NOR Flash
4 Gigabit RJ45 PoE ports
2 MT7615N wifi chips (2.4GHz and 5GHz)
2 USB ports (1xUSB2 and 1xUSB3 - GL3510 chip)
RJ45 RS232 port on front panel (Max3232 chip)
2x mPCIe 2.0 slots for 4G/5G cards
2x SIM slot
1x SDCard Slot
Power via DC12V
4x Cell Antennae
4x Wifi Antennae

MAC Address Locations:
Purpose	Ex.		Partition	Offset
2.4 Ghz	*:01	factory		0x4
5 GHz	*:02	factory		0x8004
LAN		*:03	factory		0xe000
WAN		*:04	factory		0xe006

MAC address prefix E4:3A:65 is registered to MofiNetwork Inc and used as the prefix for all MAC addresses.

Manual: https://mofinetwork.com/files/MoFi_Network_MOFI5500_5GXeLTE_EM7690_SPECS.pdf
WiFi chip specs: https://www.mediatek.com/products/broadband-wifi/mt7615
CPU chip specs: https://www.mediatek.com/products/home-networking/mt7621
Teardown Pictures: https://fccid.io/2AE6X-MOFI5500/Internal-Photos/Internal-Photos-5591739

Installation:

Update Mofi 5500 to at least stock firmware version 4.8.6. (Available on the Mofi website.) Previous versions are untested in the upgrade process. Log into the LuCI web interface, usually at 192.168.10.1 and visit the 'System->Backup/Flash Firmware' page. Upload and flash the firmware as usual.

Note to Maintainers: Do not remove SUPPORTED_DEVICES from the Makefile! The customized Mofi version of OpenWRT (stock firmware) expects to see mofi5500 as the device name. The stock firmware does not allow for forcing an installation. Without this line, users cannot upload the new firmware through the stock Mofi firmware.

This device uses cell modems that could use QMI or MBIM. Add LuCI Modem Manager to allow people to use these. Also, if they have two cell network cards, ethernet, USB, or other kinds of networks, they may wish to use MWAN3 to allow failover amongst their networks. Please compile it with mwan3 for multiple WAN connections.

Co-authored-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Signed-off-by: Rick Mac Gillis <noreply@rickmacgillis.com>

blocktrron commented:
Merged to my staging tree, thanks!
